### PR TITLE
Improve rule structure, organize code and add more rules

### PR DESCRIPTION
**Pull request by PCXDME**
#51

### CODE REVIEW

File: api/nlp/nlp.js [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+function getActions(sentence, chartType, data, layout) {
+  const nlpSentence = nlp(sentence);
+
+  if (!sentenceRules[chartType] || !tagRules[chartType] || !chartType || !data || !layout) {
+    return null;
+  }
+
+  // Try matching general rules first then try matching chart type specific rules.
+  return (
+    getActionsBySentenceRuleList(sentenceRules.general, nlpSentence, data, layout) ||
```
Comment (fhnaseer) [inline api/nlp/nlp.js:73]: Can we stick to only one? Either use sentenceRules or TagRules?,

File: api/controllers/nlpController.js [hunk 2]
```diff
@@ -1,87 +1,10 @@
...
-  return (
-    getActionsBySentenceRuleList(sentenceRules.general, nlpSentence, data, layout) ||
-    getActionsBySentenceRuleList(sentenceRules[chartType], nlpSentence, data, layout) ||
-    getActionsByTagRuleList(tagRules.general, nlpSentence, data, layout) ||
-    getActionsByTagRuleList(tagRules[chartType], nlpSentence, data, layout)
-  );
-}
+const NLP = require('../nlp/nlp');
 
 function postRequest(req, res) {
```
Comment (fhnaseer) [inline api/controllers/nlpController.js]: Add comments for this function,

Comment (PCXDME) [reply api/controllers/nlpController.js]: Added

**Review by fhnaseer — APPROVED** [review]

File: api/nlp/nlp.js (hunk 1 above)
Comment (PCXDME) [reply api/nlp/nlp.js:73]: Tags are simpler and easier to match. However, some sentence are hard to be match only by tags, so we need sentence rules. The system will try to match the sentence with sentence rules first and then try to match tag rules.
The sentence rules will be similar to regex, but for compromise.
The tag rules will match based on tags of the sentence.

**Comment by fhnaseer** [comment]
Can you do this and add it in this merge request?,
group 4, when ever an api call fails we receive a 400 bad request error with no message. I would suggest to implement proper error messages for better debugging.. Thanks !!

**Comment by PCXDME** [comment]
OK

**Review by fhnaseer — APPROVED** [review]

**Pull request merged**